microcosm/config/model.py

```diff
@@ -96,12 +143,14 @@ def validate(self, metadata, path, value):
                 value = self.mock_value
             elif self.default_value is not None:
                 value = self.default_value
+            elif self.default_factory is not None:
+                value = self.default_factory()
             elif not value.required:
                 return None
             else:
                 raise ValidationError(f"Missing required configuration for: {'.'.join(path)}")
 
         try:
             return self.type(value)
-        except ValueError:
+        except (ValueError, TypeError):
```

Comment:
We want to make sure we fail when they provide `None` to something that doesn't expect `None`